src/log/SubsystemMap.h

```diff
@@ -6,40 +6,54 @@
 
 #include <string>
 #include <vector>
+#include <algorithm>
+
+#include "common/subsys_types.h"
 
 #include "include/assert.h"
 
 namespace ceph {
 namespace logging {
 
-struct Subsystem {
-  int log_level, gather_level;
-  std::string name;
-  
-  Subsystem() : log_level(0), gather_level(0) {}     
-};
-
 class SubsystemMap {
-  std::vector<Subsystem> m_subsys;
-  unsigned m_max_name_len;
+  // Access the current gathering levels must be *FAST* as they are
+  // read over and over from all places in the code (via should_gather()
+  // by i.e. dout).
+  std::array<uint8_t, ceph_subsys_get_num()> m_gather_levels;
+
+  // The rest. Should be as small as possible to unnecessarily not
+  // enlarge md_config_t and spread it other elements across cache
+  // lines. Access can be slow.
+  std::vector<ceph_subsys_item_t> m_subsys;
```

Comment:
No need to carry the size as its already known. `std::unique_ptr<std::array>` instead?